### PR TITLE
[admin] Added notification widget

### LINKED ISSUE #4: [admin] Add a JS widget to browse notifications
- [x] The widget should support infinite scrolling.
- [x] User should be able to read description of notifications.
- [x] Opening a notification should mark them as read

**Project Idea**
> Create a javscript widget for the admin which allows to view all notification and expand their details by clicking on it.
This widget shall support infinite scrolling (we have a nice implementation in django-ipam which can be reused, either partially or totally) and expand the markdown into HTML.
The operation of opening a notification detail shall flag it as read.
This feature will require adding API endpoints to get notification list and details.

Should be done after #3

### LINKED ISSUE #12: [websockets] Delivering notifications at real-time
Notifications should be received and displayed on admin dashboard without reloading the page. A message(toast) should also be to notify upon receiving a new notification.

### PR DESCRIPTION
**Pull request by pandafy**
Added JS widget for browsing notifications
Added infinite scroll in JS widget
Added websockets for updating notifications
Added notification toast message for real-time notifications using websockets
Added notification sound which is played on toast notification
Removed NotificationAdmin

Closes #12
Closes #4

### CODE REVIEW

**Comment by NoumbissiValere** [comment]
> Hey @pandafy @Vivekrajput20 @NoumbissiValere, I think the JS widget an this work are closely related.
> 
> Wouldn't it be easier for you @pandafy if we worked on both in one PR only?
> Otherwise we'll need to do back and forth from one PR to the other, rebasing/merging changes from one review to this other PR.
> 
> Wouldn't it be simpler, in this case, if we all worked in one branch and one PR?
> 
> Let me know what you think.

I think so too. This will greatly ease testing.

File: runtests.py [hunk 1]
```diff
@@ -17,3 +18,4 @@
     else:
         args.insert(2, 'openwisp2')
     execute_from_command_line(args)
+    sys.exit(pytest.main(['openwisp_notifications/tests/test_websockets.py']))
```
Comment (nemesisdesign) [inline runtests.py:21]: @atb00ker look at this :point_up: 
Should be useful for: https://github.com/openwisp/openwisp-controller/issues/223

**Comment by pandafy** [comment]
> whenever I click on a notification, I get redirected to something like http://example.com/admin/config/device/cc5c1d61-8e0b-4f31-9ab4-70cf6bb0c55b/change/. I ensured that the site content in the DB is updated but I still notice this behavior, do you know why this happen?

Django "Sites" caches domain on first call and subsequent calls use data stored in that cache. Please restart the server after making changes in DB. :smile: 

From the [documentation](https://docs.djangoproject.com/en/3.0/ref/contrib/sites/#caching-the-current-site-object):

>As the current site is stored in the database, each call to Site.objects.get_current() could result in a database query. But Django is a little cleverer than that: on the first request, the current site is cached, and any subsequent call returns the cached data instead of hitting the database.

File: openwisp_notifications/static/openwisp_notifications/images/icons/README.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Following icons are taken from `Font Awesome <https://fontawesome.com/>`_ project:
```
[the remark this answers is not in the document]
Comment (pandafy) [reply openwisp_notifications/static/openwisp_notifications/images/icons/README.rst:1]: This ways feels more cleaner to me. :sweat_smile:

[the remark this answers is not in the document]
Comment (nemesisdesign) [reply openwisp_notifications/static/openwisp_notifications/images/icons/README.rst:1]: it's clean but it's hidden, nobody will read it and hence we're not really giving credit in an honest way, I don't like when others do this with my work and I don't want to do it to them. The text will need to be adjusted to point out the path to the icons.

**Comment by pandafy** [comment]
> The icon doesn't look aligned correctly, I get this on both FF and Chrome.

it was shown as intended. Maybe the way I pictured it was obsolete. I have modified as below and I think it makes more sense now. :smile: 

Screenshot from 2020-07-15 23-44-45

> When narrowing at max, I get this:

It was breaking at non standard media query point, that's why I didn't took into account. I have made changes now to handle it. 

Fixed loader too and tested it on Firefox and Chromium. :grimacing:

**Comment by pandafy** [comment]
> @pandafy please implement cache invalidation, we can't ask users to restart openwisp for every change they do to some cached object in the DB.
> Connect a receiver to post_save and post_delete of Site and use `Site.objects.clear_cache()`

It is already done in Django code, I wonder why it is not working. 

https://github.com/django/django/blob/156a2138db20abc89933121e4ff2ee2ce56a173a/django/contrib/sites/models.py#L103-L120

**Comment by pandafy** [comment]
I purposely misaligned icon, I will change it. I thought it will give it a better look. (Motivation from StackOverflow's)

**Comment by pandafy** [comment]
I have tweaked CSS to reflect requested changes. :smile: 
Increasing padding more than 6px made notification level icon and text look detached. I think it is enough.

Screenshot from 2020-07-18 01-57-25

I tried replicating Django Sites bug, but I could not. I have attached a GIF demonstrating the steps I took. If I did something wrong please let me know. 


Hnet-image (1)

I refreshed web page after restarting development server.

**Comment by pandafy** [comment]
> Also, can you please point out where in this patch is the caching invalidation of the `Site` objects being handled?

This patch does not handle cache invalidation for "Site" object. Django itself handles it. [Please see the code in Django's source code](https://github.com/django/django/blob/156a2138db20abc89933121e4ff2ee2ce56a173a/django/contrib/sites/models.py#L103-L120).

The cache was not getting invalidated in openwisp-monitoring because `SITE_ID` was not configured properly in `settings.py`. 
It was configured as
```
SITE_ID = '1'
```
instead of
```
SITE_ID = 1
```
i.e. `str` instead of `int`. 

I have made this change in [openwisp-monitoring PR](https://github.com/openwisp/openwisp-monitoring/pull/158/files#diff-9dec27df1ed1039b50d2b96520f1de08R80). It is now working as expected. I think this is the reason why it was working properly here but not in openwisp-monitoring.

**Comment by pandafy** [comment]
Updated CSS :smile: 

Screenshot from 2020-07-18 17-21-07

**Pull request merged**